src/Microsoft.Restier.Publishers.OData/Model/RestierModelBuilder.cs

```diff
@@ -49,7 +55,19 @@ class RestierModelBuilder : IModelBuilder
 
             context.EntitySetTypeMapCollection.Clear();
 
+            ExtendModel(builder);
+
             return builder.GetEdmModel();
         }
+
+        /// <summary>
+        /// This method allow user to extend the model with more entity type and entity set before operation model builder starts
+        /// A sub class need to be implemented and register as DI service like "AddTransient&lt;RestierModelBuilder, CustomizedRestierModelBuilder&gt;"
+        /// </summary>
+        /// <param name="builder"></param>
+        public virtual void ExtendModel(ODataConventionModelBuilder builder)
```

Comment:
@advancedrei thanks for your great comments, first let's see what we want to achieve,

User need to build a model, and here are what we need to build,
1. Some entity sets and entity types and these are from DbContext (For some user who does not need this part, he will not have this part). (Done by RestierModelBuilder with data set in ModelContext by EfModelBuilder)
2. More entity sets and entity types been built besides the ones from DbContext (small group of consumers) (User need to add something here, this is what we simplify with this PR)
3. Entity sets and types built by end user who does not use EF (must for people who does not use EF). (Already support, as inner most service)
4. Restier build additional entity sets, singletons defined in Api class (done by RestierModelExtender) 
5. Restier build operation model directly from method defined in Api class (done by RestierOperationModelBuilder).
6. User may customize/log/add more element to the model (Already support, as out most service)

Also we want to allow user to use any way to build model which means user can build with OData lib Edm class directly or with Web Api OData conversion model build.

Then let discuss the design point,
1. For users who use EF, we expect they do not need to touch model build at all, all are done by restier model builder.
2. For users who need additional entity sets and entity types or who does not use EF (refer to item 2), we allow them to extend, but we want to be extend more easily, then his does not need to touch operation model build.
3. For users who may customize/log/add more element to the model, we allow him to customize and most he need to do with model directly like set a new annotation.

So we plan to return model directly, but user can create a new conversion model builder with existing model, we open issue https://github.com/OData/WebApi/issues/755

You provide a great option to achieve this, we may extract existing logic into different methods and user can call these method in some specified orders. We will have more discussions on this part.